aws/s3.go

```diff
@@ -0,0 +1,159 @@
+			}
+		} else {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+// getAllS3Buckets lists and returns a map of per region AWS S3 buckets which were created before excludeAfter
+func getAllS3Buckets(session *session.Session, excludeAfter time.Time) (map[string][]*string, error) {
+	svc := s3.New(session)
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+	for _, bucket := range output.Buckets {
+		if !excludeAfter.After(*bucket.CreationDate) {
+			continue
+		}
+
+		input := &s3.GetBucketLocationInput{
+			Bucket: aws.String(*bucket.Name),
+		}
+
+		result, err := svc.GetBucketLocation(input)
+		if err != nil {
+			// It might be possible that between listing the bucket and finding the location - the target bucket got deleted
+			// Verify if that is the case.
+			exists, err := checkS3BucketExists(svc, *bucket.Name)
+			if err != nil {
+				return nil, errors.WithStackTrace(err)
+			}
+			if !exists {
+				logging.Logger.Warnf("Failed to get bucket location for - %s - ignoring as bucket might be deleted.", *bucket.Name)
+				continue
+			}
+		}
+
+		var bucketRegion string
+		if result.LocationConstraint == nil {
+			// GetBucketLocation returns nil for us-east-1
+			// https://github.com/aws/aws-sdk-go/issues/1687
+			bucketRegion = "us-east-1"
+		} else {
+			bucketRegion = *result.LocationConstraint
+		}
+
+		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
+			bucketNamesPerRegion[bucketRegion] = []*string{}
+		}
+
+		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+	}
+
+	return bucketNamesPerRegion, nil
+}
+
+// nukeAllS3Buckets deletes all S3 buckets passed as input
+func nukeAllS3Buckets(session *session.Session, bucketNames []*string) (delCount int, err error) {
+	svc := s3.New(session)
+
+	if len(bucketNames) == 0 {
+		logging.Logger.Infof("No S3 Buckets to nuke in region %s", *session.Config.Region)
+		return 0, nil
+	}
+
+	bucketCount := 0
+	totalCount := len(bucketNames)
+
+	logging.Logger.Infof("Deleting - %d S3 Buckets in region %s", totalCount, *session.Config.Region)
+
+	for _, bucketName := range bucketNames {
+		// It might be possible that between listing the bucket and deleting the bucket - the target bucket got deleted
+		// Verify if that is the case.
+		exists, err := checkS3BucketExists(svc, *bucketName)
+		if err != nil {
+			return delCount, err
+		}
+		if !exists {
+			logging.Logger.Warnf("Failed to find bucket - %s - ignoring as bucket might be deleted.", *bucketName)
+			continue
+		}
+
+		bucketCount++
+		logging.Logger.Infof("Deleting - %d/%d - S3 Bucket: %s", bucketCount, totalCount, *bucketName)
+
+		iter := s3manager.NewDeleteListIterator(svc, &s3.ListObjectsInput{
+			Bucket: bucketName,
+		})
+
+		if err := s3manager.NewBatchDeleteWithClient(svc).Delete(aws.BackgroundContext(), iter); err != nil {
```

Comment:
By the name, it sounds like `BatchDelete` is used to delete multiple buckets, but it looks like we're using it to delete one at a time? Could we pass it the whole list, broken down into batches, instead?

Reply:
I was confused too, and then I looked up the docs and it turns out this API deletes all objects that match the filter passed in to the iterator, where the filter is `ListObjectsInput`: listing objects of a bucket.

Either way I think a comment before this section, or wrapping this in a function named `emptyBucket` would make it clearer.

Reply:
Understood. Yes BatchDelete was to take care of non-empty buckets. Will call it nukeS3BucketObjects for consistentcy.